prototype/tests/test_models.py

```diff
@@ -67,3 +73,142 @@ def test_case_versioning(self, testapp):
+                        attachments=attachments or None)
+            )
+
+        db.session.bulk_save_objects(forms)
+
+    @pytest.mark.skipif(True, reason="Not a real test. Only run this manually")
+    def test_table_size_case(self):
+        """
+        Insert 10000 rows into the form_data table. Run scripts/get_table_sizes.sh to
+        output the resulting DB size.
+        """
+        # Config that varies the row size
+        case_properties_per_row = 75
+        forms_per_case = 10
+        attachments_per_row = 1
+        domain = 'test-domain'
+
+        num_rows = 10000
+
+        delete_all_data()
+        cases = []
+        for i in range(num_rows):
+            attachments = {
+                _random_string(20): {
+                    'mime': random.choice(ATTACHMENT_TYPES.values()),
+                    'size': 5000,
+                    'id': _random_string(20),
+                    'src': _random_string(50),
+                    'from': 'local',
+                    'name': _random_string(20),
+                    'md5': hashlib.md5(str(uuid4())).hexdigest(),
+                    'props': {'w': 240, 'h': 164}
+                }
+                for i in range(attachments_per_row)
+            }
+
+            case_json = case_json_compact(case_properties_per_row, forms_per_case)
+            cases.append(
+                CaseData(id=str(uuid4()), domain=domain, owner_id=str(uuid4()),
+                        server_modified_on=datetime.utcnow(),
+                        case_json=case_json,
+                        attachments=attachments or None)
+            )
+
+        db.session.bulk_save_objects(cases)
+
+
+def case_json_expanded(num_props, num_forms):
+    """
+    {
+        'property_name': {
+            'form_id': 'ID of the form that last touched this property',
+            'value': 'property value',
+            'date': 'date of last update'
+        },
+        ...
+    }
+    """
+    form_ids = [str(uuid4()) for i in range(num_forms)]
+    case_json = {
+        _random_string(20): {
+            'form_id': random.choice(form_ids)[0],
+            'value': _random_string(20),
+            'date': json_format_datetime(datetime.utcnow())
+        }
+        for i in range(num_props)
+    }
+    return case_json
+
+
+def case_json_compact(num_props, num_forms):
+    """The goal of the compact format is to reduce the overall size of the JSON
+    by removing duplicated data. Mostly this is form IDs and form dates.
+
+    Compact format look like this:
+    {
+        'f': {
+            'unique_form_id_prefix': [form_id, form_date]
+            ...
+        },
+        'p': {
+            'property_name': [form_id_prefix, value]
+            ...
+        }
+    }
+
+    The 'unique_form_id_prefix' is as many characters as are needed from the front of the form_id
+    in order to make it unique in the 'f' dict.
```

Comment:
i'm still not 100% sure we need to store any of this if we only support rebuildng cases from forms (and not from themselves)